### DISCUSSION TITLE
File uploading directly from Node JS

### DISCUSSION THREAD

**Discussion by lakefox** [pocketbase/pocketbase#1201] [Q&A] [answered]
I've been trying to get file uploading working with the js-sdk on node instead of in the browser. Is this at all possible? To supplement FormData I am using the NPM package "form-data" which I think should work but isn't currently. Here is what I have so far.
```javascript
// Log the user in using a authorization header
pb.authStore.loadFromCookie(req.headers.authorization);

// Convert the file data from a POST request to a buffer
let fileBuffer = Buffer.from(req.body.data, "binary");

let form = new FormData();

// Set fields
form.append("username", req.body.username);
form.append("name", req.body.name);
form.append("file", fileBuffer);

await pb.collection('assets').create(form);
// This always ends in the same error "ClientResponseError 0"
```

**Comment by ganigeorgiev** [comment, marked as answer]
_This seems similar to https://github.com/pocketbase/pocketbase/discussions/278#discussioncomment-3356038._

`FormData` is natively supported in Node 18 and I recommend updating and using the native implementation since it comes also with support for other web apis (eg. `fetch`).

`FormData` by default works only with string and `Blob` for field values (https://developer.mozilla.org/en-US/docs/Web/API/FormData/append), so you can't pass Buffer and this is probably supported only by the library you are using, but when sending the request `fetch` expects spec compliant `FormData` format so that it can construct the correct multipart header and body.

I'm not sure how `req.body` was parsed and what is the format of `req.body.data`, but you can try loading it into a `Blob`, eg:

```js
formData.append("file", new Blob([req.body.data]), "test.png");
```

If it is a local/temp file path, you may need to read it first, aka:

```js
import fs from "fs";

formData.append("file", new Blob([fs.readFileSync(req.body.data)]), "test.png");
```

**Reply by lakefox** [reply]
Thank you so much, removing form-data and using the native implementation fixed it.

**Comment by tteke** [comment]
Is there a way to use streams in this or does native implementation of formdata not support streams at all ? Since reading a whole file with readFileSync is basically writing the whole file to the memory (I think?). In that case small files should be fine but large files might cause issues, it feels like.

**Reply by ganigeorgiev** [reply]
The JS SDK uses whatever global `fetch` and respectively `FormData` implementations are available.

By spec the `FormData` field value could be either a string or Blob (including subclasses such as [File](https://developer.mozilla.org/en-US/docs/Web/API/File)). Everything else is stringified.

Technically `Blob` could be backed by a stream but this depends on the runtime interface impementation and how the `fetch` will operate with it. I'm not that familiar with the Node.js APIs so I'm not sure, but from quick search there seems to be experimental [`fs.openAsBlob` method](https://nodejs.org/api/fs.html#fsopenasblobpath-options) that you can give a try.

But again, there is nothing really PocketBase specific here. It should work as long as the submitted request is valid `multipart/form-data`.

**Reply by tteke** [reply]
Yes of course its not pocketbase specific issue, just wanted to chip in on the discussion cuz of my curiosity, thanks for the explanation.
